daemon/containerd/image_tag.go

```diff
@@ -50,11 +50,14 @@ func (i *ImageService) TagImageWithReference(ctx context.Context, imageID image.
 		}
 
 		// If there already exists an image with this tag, delete it
-		if err := i.softImageDelete(replacedImg); err != nil {
+		if err := i.softImageDelete(ctx, replacedImg); err != nil {
 			logger.WithError(err).Error("failed to delete old image")
 			return errdefs.System(err)
 		}
 
+		// From this point ignore context cancellation.
+		ctx = context.Background()
```

Comment:
The slightly confusing bit here is that;

- If we didn't get an error on the first `img.Create()`, we get in this "if" branch, and replace the `ctx`  with new context
- However, the `sourceDanglingImg, err := is.Get(ctx, danglingImageName(target.Digest))` also uses the context, so in one case it uses the original context, but in another case it uses a new `context.Background()`.


Should the second (`sourceDanglingImg ... `) code have a new context depending on this? If that also should always be executed with a separate context, perhaps the `context.Background()` should be inlined in both cases (no context shared between the two).

Do any of those contexts need a timeout, or not needed?

Reply:
Right, the code starting from `sourceDanglingImg...` shouldn't depend on the inner code from the previous if branch, but still, if we want to have this "uninterruptible" behavior, then it should independently use other non-cancellable context.

An independent timeout could be a good idea indeed. Just in case there's some unexpected error (grpc/containerd issues?).